File: app/schemas/command_center.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import (
    DecisionCategory,
    EngineeringChangeStatus,
    ExecutionStatus,
    ProjectHealth,
    ProjectPriority,
    ProjectRiskType,
    ProjectStage,
    TimelineStepStatus,
)
from app.schemas.common import TimestampSchema
from app.schemas.project import ProjectRead
from app.schemas.timesheet import TimesheetEntryRead


class ProjectFolderPaths(BaseModel):
    project_folder_path: str | None = None
    cad_folder_path: str | None = None
    released_folder_path: str | None = None
    suggested_project_folder: str | None = None
    suggested_cad_folder: str | None = None
    suggested_released_folder: str | None = None


class ProjectFolderPathsUpdate(BaseModel):
    project_folder_path: str | None = Field(default=None, max_length=500)
    cad_folder_path: str | None = Field(default=None, max_length=500)
    released_folder_path: str | None = Field(default=None, max_length=500)


class TimelineStep(BaseModel):
    milestone_id: UUID | None = None
    name: str
    sort_order: int
    status: TimelineStepStatus
    due_date: date | None = None
    completed_at: datetime | None = None


class ProjectKpis(BaseModel):
    completion_percent: Decimal
    quoted_hours: Decimal
    actual_hours: Decimal
    remaining_hours: Decimal
    variance: Decimal
    variance_percent: Decimal
    budget_consumption_percent: Decimal
    days_remaining: int
    current_milestone: str | None = None


class TeamMemberCapacity(BaseModel):
    user_id: UUID
    name: str
    role: str
    capacity_hours: Decimal
    allocated_hours: Decimal
    available_hours: Decimal
    availability_status: str


class ProjectTeamSummary(BaseModel):
    engineering_manager_name: str | None = None
    design_leader_name: str | None = None
    designer_name: str | None = None
    surfacer_name: str | None = None
    team_name: str | None = None
    team_colour: str | None = None
    members: list[TeamMemberCapacity] = Field(default_factory=list)


class CustomerProjectSummary(BaseModel):
    customer_id: UUID
    customer_name: str
    customer_code: str | None = None
    primary_contact_name: str | None = None
    primary_contact_email: str | None = None
    active_projects: int = 0
    completed_projects: int = 0
    average_hours: Decimal = Decimal("0")


class ProjectDecisionRead(TimestampSchema):
    id: UUID
    project_id: UUID
    user_id: UUID
    user_name: str | None = None
    category: DecisionCategory
    comment: str
    milestone_id: UUID | None = None
    milestone_name: str | None = None


class ProjectDecisionCreate(BaseModel):
    category: DecisionCategory = DecisionCategory.general
    comment: str = Field(min_length=1)
    milestone_id: UUID | None = None


class ProjectDecisionUpdate(BaseModel):
    category: DecisionCategory | None = None
    comment: str | None = Field(default=None, min_length=1)
    milestone_id: UUID | None = None


class EngineeringChangeRead(TimestampSchema):
    id: UUID
    project_id: UUID
    ec_number: str
    title: str
    status: EngineeringChangeStatus
    hours: Decimal
    closed_at: datetime | None = None


class EngineeringChangeCreate(BaseModel):
    ec_number: str = Field(max_length=50)
    title: str = Field(max_length=255)
    hours: Decimal = Field(default=Decimal("0"), ge=0)


class EngineeringChangeUpdate(BaseModel):
    ec_number: str | None = Field(default=None, max_length=50)
    title: str | None = Field(default=None, max_length=255)
    status: EngineeringChangeStatus | None = None
    hours: Decimal | None = Field(default=None, ge=0)


class EngineeringChangeSummary(BaseModel):
    open_count: int = 0
    closed_count: int = 0
    total_hours: Decimal = Decimal("0")
    items: list[EngineeringChangeRead] = Field(default_factory=list)


class ProjectRiskItem(BaseModel):
    risk_type: ProjectRiskType
    severity: str
    title: str
    detail: str | None = None


class CommandCenterHeader(BaseModel):
    tool_number: str
    part_description: str
    customer_name: str
    team_name: str | None = None
    designer_name: str | None = None
    project_stage: ProjectStage
    execution_status: ExecutionStatus
    current_milestone: str | None = None
    completion_percent: Decimal
    health: ProjectHealth
    priority: ProjectPriority
    days_remaining: int


class ProjectCommandCenter(BaseModel):
    project: ProjectRead
    header: CommandCenterHeader
    timeline: list[TimelineStep] = Field(default_factory=list)
    kpis: ProjectKpis
    team: ProjectTeamSummary
    customer_summary: CustomerProjectSummary
    decisions: list[ProjectDecisionRead] = Field(default_factory=list)
    recent_timesheets: list[TimesheetEntryRead] = Field(default_factory=list)
    engineering_changes: EngineeringChangeSummary
    risks: list[ProjectRiskItem] = Field(default_factory=list)
    folders: ProjectFolderPaths
    hours: dict[str, Decimal] = Field(default_factory=dict)
    milestone_summary: dict[str, int | Decimal] = Field(default_factory=dict)

File: frontend/src/types/CommandCenter.ts
import type { ExecutionStatus, ProjectHealth, ProjectStage } from './common';
import type { Project } from './Project';
import type { TimesheetEntry } from './TimesheetEntry';

export type TimelineStepStatus = 'completed' | 'current' | 'upcoming' | 'delayed';
export type DecisionCategory =
  | 'design'
  | 'customer'
  | 'manufacturing'
  | 'tooling'
  | 'schedule'
  | 'quality'
  | 'general';
export type ProjectRiskType =
  | 'milestone_delay'
  | 'hours_over_quote'
  | 'overdue'
  | 'missing_approvals'
  | 'designer_overloaded';

export interface CommandCenterHeader {
  tool_number: string;
  part_description: string;
  customer_name: string;
  team_name: string | null;
  designer_name: string | null;
  project_stage: ProjectStage;
  execution_status: ExecutionStatus;
  current_milestone: string | null;
  completion_percent: number;
  health: ProjectHealth;
  priority: 'critical' | 'high' | 'medium' | 'low';
  days_remaining: number;
}

export interface TimelineStep {
  milestone_id: string | null;
  name: string;
  sort_order: number;
  status: TimelineStepStatus;
  due_date: string | null;
  completed_at: string | null;
}

export interface ProjectKpis {
  completion_percent: number;
  quoted_hours: number | null;
  actual_hours: number;
  remaining_hours: number | null;
  variance: number | null;
  variance_percent: number | null;
  budget_consumption_percent: number | null;
  days_remaining: number;
  current_milestone: string | null;
  working_model_id?: string | null;
  working_model_code?: string | null;
  working_model_name?: string | null;
  strategy_key?: string | null;
  show_quoted_variance?: boolean;
  show_over_budget_indicators?: boolean;
  model_metrics?: Record<string, number | string | null>;
}

export interface TeamMemberCapacity {
  user_id: string;
  name: string;
  role: string;
  capacity_hours: number;
  allocated_hours: number;
  available_hours: number;
  availability_status: string;
}

export interface ProjectTeamSummary {
  engineering_manager_name: string | null;
  design_leader_name: string | null;
  designer_name: string | null;
  surfacer_name: string | null;
  team_name: string | null;
  team_colour: string | null;
  members: TeamMemberCapacity[];
}

export interface CustomerProjectSummary {
  customer_id: string;
  customer_name: string;
  customer_code: string | null;
  primary_contact_name: string | null;
  primary_contact_email: string | null;
  active_projects: number;
  completed_projects: number;
  average_hours: number;
}

export interface ProjectDecision {
  id: string;
  project_id: string;
  user_id: string;
  user_name: string | null;
  category: DecisionCategory;
  comment: string;
  milestone_id: string | null;
  milestone_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface EngineeringChange {
  id: string;
  project_id: string;
  ec_number: string;
  title: string;
  status: 'open' | 'closed';
  hours: number;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface EngineeringChangeSummary {
  open_count: number;
  closed_count: number;
  total_hours: number;
  items: EngineeringChange[];
}

export interface ProjectRiskItem {
  risk_type: ProjectRiskType;
  severity: string;
  title: string;
  detail: string | null;
}

export interface ProjectFolderPaths {
  project_folder_path: string | null;
  cad_folder_path: string | null;
  released_folder_path: string | null;
  suggested_project_folder: string | null;
  suggested_cad_folder: string | null;
  suggested_released_folder: string | null;
}

export interface ProjectCommandCenter {
  project: Project;
  header: CommandCenterHeader;
  timeline: TimelineStep[];
  kpis: ProjectKpis;
  team: ProjectTeamSummary;
  customer_summary: CustomerProjectSummary;
  decisions: ProjectDecision[];
  recent_timesheets: TimesheetEntry[];
  engineering_changes: EngineeringChangeSummary;
  risks: ProjectRiskItem[];
  folders: ProjectFolderPaths;
  hours: Record<string, number>;
  milestone_summary: Record<string, number>;
}
